promql/engine.go

```diff
@@ -1077,11 +1129,50 @@ func (ev *evaluator) eval(expr Expr) Value {
 			panic(fmt.Errorf("cannot do range evaluation of matrix selector"))
 		}
 		return ev.matrixSelector(e)
+
+	case *SubqueryExpr:
+		offsetMillis := durationToInt64Millis(e.Offset)
+		rangeMillis := durationToInt64Millis(e.Range)
+		newEv := &evaluator{
+			endTimestamp:   ev.endTimestamp - offsetMillis,
+			ctx:            ev.ctx,
+			currentSamples: ev.currentSamples,
+			maxSamples:     ev.maxSamples,
+			logger:         ev.logger,
+		}
+
+		newEv.interval = durationToInt64Millis(e.Step)
+		if newEv.interval == 0 {
+			defaultEvaluationIntervalMtx.RLock()
+			newEv.interval = durationToInt64Millis(DefaultEvaluationInterval)
+			defaultEvaluationIntervalMtx.RUnlock()
+		}
+
+		// We want to align the start time of the subquery with its step.
+		// Aligned steps are `0 1*step 2*step 3*step ...`
+		// We choose start time as the first aligned step just after (or equal to)
+		// 'S = (ev.startTimestamp - offset - range)'
+
+		// numSteps = S / step
+		// newEv.startTimestamp = step * numSteps
+		newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
```

Comment:
You can get rid of the `if` below if you write this as:
```suggestion
		newEv.startTimestamp = ((ev.startTimestamp - offsetMillis - rangeMillis - 1) / newEv.interval + 1) * newEv.interval
```

I would also replace all the comments on the few lines above with "Start from the first timestamp at or after \`ev.startTimestamp - offset - range\` that is a multiple of \`interval\`.

Reply:
I was testing this, and noticed that when the start time goes negative because of offset and range, this equation wont hold good.